### PR TITLE
Add implementation

### PR DESCRIPTION
**Pull request by remcohaszing**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

This adds the initial implementation of the package.

The idea comes from https://github.com/syntax-tree/hast-util-from-html/issues/2

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    "xo": "^0.53.0"
+  },
+  "scripts": {
+    "prepack": "npm run build && npm run format",
+    "build": "tsc --build --clean && tsc --build && type-coverage",
+    "format": "remark . -qfo && prettier . -w --loglevel warn && xo --fix",
+    "test-browser": "node --conditions development --conditions browser --test",
+    "test-node": "node --conditions development --test",
+    "test-worker": "node --conditions development --conditions worker --test",
+    "test-api": "npm run test-browser && npm run test-node && npm run test-worker",
```
Comment (remcohaszing) [inline package.json:67]: This runs the same tests against the different export conditions.

File: readme.md
Comment (remcohaszing) [inline readme.md:1]: This is basically a modified version of the readme from [`hast-util-from-html`](https://github.com/syntax-tree/hast-util-from-html).

File: lib/browser.js
Comment (remcohaszing) [inline lib/browser.js:1]: I tried to match the output of [`hast-util-from-html`](https://github.com/syntax-tree/hast-util-from-html) here. I’m not entirely familiar with the details though. I though it’s about right to parse the document using a `<template>` element in fragment mode, and using a dom parser otherwise.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+/**
+ * @typedef {import('./lib/index.js').Options} Options
+ */
+
+export {fromHtml} from './lib/index.js'
```
Comment (wooorm) [inline index.js]: I’d prefer `fromHtmlIsomorphic`, to match the package name without prefix?

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/**
+ * @typedef {import('hast').Root} Root
+ * @typedef {Pick<import('hast-util-from-html').Options, 'fragment'>} Options
+ */
+
+import {fromHtml as fromHtmlNode} from 'hast-util-from-html'
```
Comment (wooorm) [inline lib/index.js]: Different name would also make this rename no longer needed

File: package.json [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  "contributors": [
+    "Remco Haszing <remcohaszing@gmail.com>"
+  ],
+  "sideEffects": false,
+  "type": "module",
+  "exports": {
+    ".": {
+      "worker": "./index.js",
+      "browser": "./lib/browser.js",
+      "default": "./index.js"
```
Comment (wooorm) [inline package.json:34]: I think you should also include deno and react-native:
https://github.com/wooorm/decode-named-character-reference/blob/4347630741f82d7d4c7fb38763ba158951b2a1be/package.json#L34-L38.

That package depends on the DOM too, and it was really really tough to get right, so I’d feel more secure if it was included like that!

File: lib/index.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * @param {string} value
+ *   Serialized HTML to parse.
+ * @param {Options} [options]
+ *   Configuration (optional).
+ * @returns {import('hast').Root}
+ *   Tree
+ */
+export function fromHtml(value, options) {
+  return fromHtmlNode(value, options)
```
Comment (wooorm) [inline lib/index.js]: Should positional info be removed?

File: readme.md [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+[![Build][build-badge]][build]
+[![Coverage][coverage-badge]][coverage]
+[![Downloads][downloads-badge]][downloads]
+[![Size][size-badge]][size]
+[![Sponsors][sponsors-badge]][collective]
+[![Backers][backers-badge]][collective]
+[![Chat][chat-badge]][chat]
+
+[hast][] utility that turns HTML into a syntax tree.
```
Comment (wooorm) [inline readme.md]:
```suggestion
[hast][] utility that turns HTML into a syntax tree, using browser APIs
when available, so it has a smaller bundle size there.
```

Something like this to mention the main difference?

File: readme.md [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+*   [Contribute](#contribute)
+*   [License](#license)
+
+## What is this?
+
+This package is a utility that takes HTML input and turns it into a hast syntax
+tree.
+
+When running in a browser, this uses [`hast-util-from-dom`][hast-util-from-dom],
+otherwise this uses [`hast-util-from-html`][hast-util-from-html].
```
Comment (wooorm) [inline readme.md]:
```suggestion
In a browser, this uses [`hast-util-from-dom`][hast-util-from-dom],
otherwise it uses [`hast-util-from-html`][hast-util-from-html].
```

File: readme.md [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+
+## When should I use this?
+
+If you want to handle syntax trees manually in isomorphic code and care about
+bundle size, use this.
```
Comment (wooorm) [inline readme.md]:
```suggestion
If you want to get a syntax tree without positional info, and your code should
be isomorphic (it could run anywhere), as it results in a smaller bundle size.
```

File: readme.md [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+
+If you want to use positional information, use
+[`hast-util-from-html`][hast-util-from-html].
```
Comment (wooorm) [inline readme.md]:
```suggestion
If you need positional information, use
[`hast-util-from-html`][hast-util-from-html].
```

File: readme.md [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+
+If you don’t care about positional information, and your code will only run in
+the browser, use [`hast-util-from-dom`][hast-util-from-dom].
```
Comment (wooorm) [inline readme.md]:
```suggestion
If you don’t care about positional info and your code only runs in browsers,
use [`hast-util-from-dom`][hast-util-from-dom].
```

File: readme.md [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+
+Finally you can use the utility [`hast-util-to-html`][hast-util-to-html] to do
+the inverse of this utility.
+It turns hast into HTML.
```
Comment (wooorm) [inline readme.md]:
```suggestion
Finally you can use the utility [`hast-util-to-html`][hast-util-to-html],
or [`hast-util-to-dom`][hast-util-to-dom] with `.outerHTML`, to do the inverse
of this utility.
That turns hast into HTML.
```

I guess we could also make a `toHtmlIsomorphic`?

File: readme.md [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+```js
+{
+  type: 'root',
+  children: [
+    {
+      type: 'element',
+      tagName: 'h1',
+      properties: {},
+      children: [Array],
+      position: [Object]
```
Comment (wooorm) [inline readme.md]: I don’t think this should be there

File: readme.md [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+This package exports the identifier [`fromHtml`][fromhtml].
+There is no default export.
+
+### `fromHtml(value[, options])`
+
+Turn serialized HTML into a hast tree.
+
+###### Parameters
+
+<!-- To do: update link when `vfile` has new docs. -->
```
Comment (wooorm) [inline readme.md]: new docs exist! `https://github.com/vfile/vfile#compatible`

File: readme.md [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+
+Tree ([`Root`][root]).
+
+### `Options`
+
+Configuration (TypeScript type).
+
+##### Fields
+
+###### `options.fragment`
```
Comment (wooorm) [inline readme.md]:
```suggestion
###### `fragment`
```

(probably a typo on my part in the reference project)

File: readme.md [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+the browser, use [`hast-util-from-dom`][hast-util-from-dom].
+
+Finally you can use the utility [`hast-util-to-html`][hast-util-to-html] to do
+the inverse of this utility.
+It turns hast into HTML.
+
+## Install
+
+This package is [ESM only][esm].
+In Node.js (version 14.14+ and 16.0+), install with [npm][]:
```
Comment (wooorm) [inline readme.md]: We can start at Node 16+, as that’s only a month away?

File: readme.md [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+## Types
+
+This package is fully typed with [TypeScript][].
+It exports the additional type [`Options`][options].
+
+## Compatibility
+
+Projects maintained by the unified collective are compatible with all maintained
+versions of Node.js.
+As of now, that is Node.js 14.14+ and 16.0+.
```
Comment (wooorm) [inline readme.md]:
```suggestion
As of now, that is Node.js 16.0+.
```

Same as above, start at Node 16?

File: test/index.js [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+import {JSDOM} from 'jsdom'
+import {removePosition} from 'unist-util-remove-position'
+
+const jsdom = new JSDOM()
+globalThis.document = jsdom.window.document
+globalThis.DOMParser = jsdom.window.DOMParser
+
+// We use a dynamic import, so we can configure jsdom before we load the module.
+const {fromHtml} = await import('hast-util-from-html-isomorphic')
```
Comment (wooorm) [inline test/index.js]: I guess the tests only run in Node, so the browser version wouldn’t be loaded, and JSDOM isn’t needed?

**Review by wooorm — COMMENTED** [review]
Awesome that you created this! Some notes.
- Some docs stuff
- Name?
- Positional info?

File: lib/browser.js
Comment (wooorm) [reply lib/browser.js:1]: I think you can basically copy/paste this code: https://github.com/rehypejs/rehype-dom/blob/main/packages/rehype-dom-parse/lib/index.js.
That works

File: index.js (hunk 2 above)
Comment (remcohaszing) [reply index.js]: Either is fine with me. I see your point about matching the package name. I changed it.

File: package.json (hunk 4 above)
Comment (remcohaszing) [reply package.json:34]: Do Deno and React Native match the `browser` export? I would find that surprising.

I prefer to omit them, but I don’t have a very strong feeling about this.

File: lib/index.js (hunk 5 above)
Comment (remcohaszing) [reply lib/index.js]: I though about it, and I don’t know. Removing it would make the API more consistent, but with no real benefit. I can’t think of a use case where it’s explicitly desirable to omit it.

File: readme.md (hunk 11 above)
Comment (remcohaszing) [reply readme.md]: I guess we could, but I personally haven’t run into any use cases yet.

File: readme.md [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+This package exports the identifier [`fromHtmlIsomorphic`][fromhtmlisomorphic].
+There is no default export.
+
+### `fromHtmlIsomorphic(value[, options])`
+
+Turn serialized HTML into a hast tree.
+
+###### Parameters
+
+<!-- To do: update link when `vfile` has new docs. -->
```
Comment (remcohaszing) [inline readme.md]:
```suggestion
```

File: readme.md (hunk 12 above)
Comment (remcohaszing) [reply readme.md]:
```suggestion
```

File: readme.md (hunk 15 above)
Comment (remcohaszing) [reply readme.md]:
```suggestion
In Node.js (version 16.0+), install with [npm][]:
```

File: test/index.js (hunk 17 above)
Comment (remcohaszing) [reply test/index.js]: The same test is run against the various export conditions.

https://github.com/syntax-tree/hast-util-from-html-isomorphic/pull/1#discussion_r1138423272

File: lib/browser.js
Comment (remcohaszing) [reply lib/browser.js:1]: That simplied it a bit. I did keep the method using a template element and its fragment. I think that’s more elegant than creating a new DOM and using a fake fragment.

File: lib/index.js (hunk 5 above)
Comment (wooorm) [reply lib/index.js]: Removing is good for generating nodes: when you want to inject a tree into another tree, e.g., your mermaid stuff, the positional info would be all wrong!

Without the removal, users would manually have to do that when injecting after this utility, but that’s potentially a useless traversal of the tree, because in browsers there would not be positional info anyway!

File: package.json (hunk 4 above)
Comment (wooorm) [reply package.json:34]: Yep, they do. In many ways they are like browsers. In their minds I guess they report “not being Node”... :/

File: test/index.js (hunk 17 above)
Comment (wooorm) [reply test/index.js]: Ah ok, good idea.

Still kinda weird to always include a DOM? That way we’d never get a crash if we incorrectly load the browser code in Node or so? 🤔 Maybe include that as an extra file with https://nodejs.org/api/cli.html#--importmodule before the `browser` condition test? Not sure if it would work tho

File: lib/index.js (hunk 5 above)
Comment (remcohaszing) [reply lib/index.js]: Positions are now removed. :)

File: test/index.js (hunk 17 above)
Comment (wooorm) [reply test/index.js]: Oh that flag is added in V19. Maybe add a comment then to look at it again when Node 18 is EOL? 😅

File: package.json (hunk 4 above)
Comment (remcohaszing) [reply package.json:34]: Well, it’s not be great IMO, but is that’s best for compatibility, let’s add it.

File: test/index.js (hunk 17 above)
Comment (remcohaszing) [reply test/index.js]: I would also prefer to load the DOM conditionally, but I don’t know how to detect which import conditions are used.

Code coverage detects both implementations are covered anyway. It’s not the same as asserting which implementation is used by which export, but it’s something.

> Maybe include some extra code with https://nodejs.org/api/cli.html#--importmodule before the `browser` condition test?

I don’t really understand what you’re suggestion here.

Comment (remcohaszing) [reply test/index.js]: > Oh that flag is added in V19. Maybe add a comment then to look at it again when Node 18 is EOL? :sweat_smile:

Oh, that clarifies to me what you meant before. :)

Comment (remcohaszing) [reply test/index.js]: `--import` may be too new, but since JSDOM is CJS, `--require` works :smile:

**Review by wooorm — APPROVED** [review]

File: test/index.js (hunk 17 above)
Comment (wooorm) [reply test/index.js]: ah, smart!

**Comment by wooorm** [comment]
I think it looks good. Anything you’d want me to look at specifically? Anything that needs to be discussed?

**Comment by remcohaszing** [comment]
I was mostly worried there might be discrepancies between the results in Node.js / the browser regarding fragment / document mode. HTML has so many edge cases :sweat_smile:.

I think it looks good though. At least good enough for my use case.

**Comment by remcohaszing** [comment]
The repo may need some settings changes to match the test of the unified ecosystem though.

**Review by ChristianMurphy — COMMENTED** [review]
Thanks @remcohaszing!
This looks great!

**Comment by wooorm** [comment]
Released in [`1.0.0`](https://github.com/syntax-tree/hast-util-from-html-isomorphic/releases/tag/1.0.0), thank you!
